### DISCUSSION TITLE
디자인 시스템 정의에 대한 이해

### DISCUSSION THREAD

**Discussion by peanudge** [InnerCircleA/design-system#2] [General]
## 디자인 시스템 이해하기 어려운 이유

디자인 시스템 자체도 너무 범위가 크고 자료마다 디자인 시스템에서 중요시하는 부분을 다르게 다뤄서 그랬던 것 같습니다. 

<img src="https://user-images.githubusercontent.com/20085849/205556604-db205d31-ad3b-4346-825d-763a4287bcb9.png" 
 width="400" />

개발자들이 접근하는 발표는 아무래도 구현과 코드 레벨에서의 운영하는 입장이니 내용은 **자유도를 어느정도 줘야 이상적일지** 고민하는 경향이 많이 보이네요. 

아무래도 디자인 시스템을 라이브러리화해서 관리하는 팀이 따로 있을 것이고 사용하는 팀이 따로 있으니 라이브러리의 확장성, 재사용에 더욱 집중하는 모습을 많이 보였어요.

- [우아콘에서 발표했던 배시시 디자인 시스템](https://www.youtube.com/watch?v=yR1g_A1-xFs)  
   -  백오피스를 위한 디자인 시스템은 디자인 **자유도가 크게 중요하지않으니 많이 쓰이는 컴포넌트**들만 어느정도 제공해주면 이상적일 것 같다.
   - 즉, 라이브러리 인터페이스를 어느정도 노출시킬지를 고민하는 것과 일맥상통하는 것 같네요.
- [FEConf 2022 디자인 시스템, 형태를 넘어서](https://www.youtube.com/watch?v=21eiJc90ggo)
   -  Flex의 `linear` 디자인 시스템에서는 **동일한 기능을 다양한 형태로 제공해야할 사례**가 많아서 그에 대한 해답으로 Headless UI의 방식을 참조해서 유연하게 만들었다.

하지만 디자이너들이 쓴 디자인 시스템 관련 글들을 읽어보면 "디자인 철학", "디자인 가이드 라인" 등등 문서화된 것들과 지켜줘야 할 디자인 가이드 문서화에 대한 강조가 더 많았어요.

사실 위 벤다이어그램에서도 `Dev Standard` 와 `Process`는 일부에 불과하니 당연할 것일 수도 있어요.

- https://www.uxpin.com/studio/blog/design-systems-vs-pattern-libraries-vs-style-guides-whats-difference/

✅ 이런 정의를 스스로하고나서 이번 디자인 시스템을 **프론트엔드 개발자가 어떻게 바라봐야할지**를 같이 생각해보면 좋을 것 같아요.
  
# 나름의 결론

그래서 공통적으로 아래와 같이 정리했고 bold 체로 강조한 부분을 집중해서 바라보면 좋을 것 같아요.

- 화면 혹은 서비스 간에 디자인 일관성을 제공해서 사용자 경험을 높인다.
- **디자인 재사용성을 높여서 효율성을 높인다.**
- 위  두 가지를 통해 비즈니스 임팩트를 낸다. (즉, 디자인 시스템은 회사에 매우 도움된다. - 종종 도움이 될까 고민하는 상황이 있음.)
- 디자인 시스템은 **항상 진화**한다. (마치 코드와 같이, 그래서 확장성이 중요하다.)
- 사람, 회사, 직군마다 디자인 시스템은 다르게 정의될 수 있다.

다만, 개발자는 시스템화된 디자인 시스템을 어떻게 사용자(개발자)들이 어떻게 사용할지 고민을 더 해야한다고 생각합니다. **라이브러리를 어떻게 만들지 고민하는 것과 유사한 것 같아요.** 위 우아콘 배시시 영상에서도 노출 interface 수준이 핵심이었던 것처럼. 

그리고 추가로 중요시해야할 것이 다양한 사용자의 요구사항을 어느정도 제한 시켜주고 **디자인 시스템 철학**이 있어야 좋은 디자인시스템 설계가 될 것 같습니다.

이렇게 적다보니 저희가 배워가야할 것은 "디자이너들이 만든 디자인 시스템을 잘 라이브러리로 만들어서 제공하는 것"이 되겠네요.

혹시 좀 다르게 생각하는 부분이나 다른 시각으로 바로 보시면 공유해주세요~!🤔 저도 더 리서치해보고 정리해볼게요.

아무래도 생각을 글로 남기지 않으면 기억에 남지 않더라구요. 

### (고민)Foundation vs Component 기반 디자인 시스템을 나누는 기준이 뭘까요? 

여러분이 생각하는 Foundation은 어떤 기준이 될까요?

현재는 자유도가 낮다/높다 라는 말이 아직까지 추상적으로 들리는 것 같아요. 자유도는 무엇으로 측정할 수 있지?
 
#### Design System에서 언급되는 `Foundation`

디자인 시스템에서 언급되는 Foundation은 icon, typo, color, spacing 와 같이 디자인에 가장 기본인 것들을 통상적으로 말하는 것 같습니다. 

- https://gds.gmarket.co.kr/foundation

`파운데이션은 일관된 레이아웃과 그에 따른 사용자 경험을 만드는 데 필수적인 시각적 요소입니다.` 

- https://atlassian.design/foundations
```
Foundations are the visual elements needed to create engaging end-to-end user experiences. This includes guidance on iconography, typography, layout and structure.
```

- https://polaris.shopify.com/foundations
```
Polaris is the design system for the Shopify admin. It’s the shared language that guides how we build high-quality merchant experiences.
```

- https://carbondesignsystem.com/ 

(IBM 디자인 시스템은 Foundation 이 없음)

**Comment by peanudge** [comment]
자료 찾던 중 처음으로 `컴포넌트 기반의 디자인 시스템의 단점`이라고 명시적으로 나타내는 글을 발견해서 추가합니다.

> https://airbnb.design/building-a-visual-language/

## Creating the Components

Traditionally, many style guides define components as atomic components, which are then used to build more complex molecules. In theory, this works well to create coherent and flexible systems. In practice, however, what often happens is that these re-usable atoms are used many different ways, allowing all kinds of molecules to be created.
Again, **this opens the door for all kinds of disjointed experiences and makes the system harder to maintain.**

구체적인 경험이 없어 온전히 이해하기는 힘들지만 `disjointed` 라는 표현으로 단점을 나타냈는데 유추를 해보면 엄격한 제약이 있으면 컴포넌트 간의 맞지않는 부분이 있을거라는(?) 추상적인 내용이네요. 


|  |파운데이션 기반  |컴포넌트 기반|
| --- | ---- | ---|
| 자유도 | 높음 | 낮음 |
| 일관성 | 낮을 수 있음 | 높음 |

**Reply by gyutae100** [reply]
# 1. [우아콘에서 발표했던 배시시 디자인 시스템](https://www.youtube.com/watch?v=yR1g_A1-xFs)
## 개인적 견해로, `제약성(또는 일관성)`과 `유연성(또는 자유도)`은 `상충 관계`가 성립한다고 추측됩니다.
- *컴포넌트 기반의 단점은 명확*하지만, *맥락 상 디자인 시스템이 유연할 필요가 없다*면 단점이 `반드시 문제`라고 판단되지 않습니다.
- 조사 결과, 초기 잦은 요구사항 변경이 예상되는 상황에서는, 유연성을 높이면서, 동시에 제약을 낮추는 방식이 주요 방법론 같습니다.
- 반대로, 이미 명확한 사용자의 요구사항을 파악했거나, 사용성보다 생산성이 중요하거나, 이미 문제 해결을 위한 `정형화 된 패턴`이 존재하는 경우에는 `높은 제약`을 통해서 해결하는 방식 같습니다.
- 결론은 제약이 높을수록 컴포넌트의 결합 또는 유연성은 떨어질 가능성이 높지만, 맥락에 따라서는 문제가 아니라고 판단합니다.
- 예시로, 만약 과제의 최종 예정 산출물이 백오피스라면 **높은 제약성을 통해서 일관성(또는 통일성)을 성립**시키면 좋을 것 같습니다. 
- 이전에 공유해주신 `배민 백오피스`처럼, 다양한 백오피스에 일관된 사용자 경험과 개발 경험을 제공하는 경우, 특히 제약성을 통한 높은 일관성 (또는 통일성)을 확보하는 전략이 효과적인 실제 예시 같습니다.

## 결론
- 효율적인 디자인 시스템은 사용 목적에 따라 제약성과 유연성의 비중을 적절한 비중을 분배함을 의미하고, 이는 생산성을 최우선 달성 지표로 삼는다.

**Reply by peanudge** [reply]
> 배민 백오피스처럼, 다양한 백오피스에 일관된 사용자 경험과 개발 경험을 제공하는 경우, 특히 제약성을 통한 높은 일관성 (또는 통일성)을 확보하는 전략

맞아요, 배민 백오피스 발표하는 내용을 들어봤을 때 너무 많은 Backoffice의 너무 다른 스타일(네비게이션 바 등등)이 운영하는 분들의 사용성을 헤치기때문에 그랬다고 하는 것을 보면   제약이 강한 좋은 사례같아요. 다만, 백오피스는 일반적으로 **웹브라우저에서만** 많이 쓰기에 괜찮을거라는 전제가 있었던 것 같아요.

아마 [위 글에서](https://airbnb.design/building-a-visual-language/)**Airbnb 디자인 시스템에서 컴포넌트 기반**이 별로라고 이야기하는 이유는 airbnb는 전세계 사람들이 모든 플랫폼(안드로이드, iOS, 테블릿, 웹 etc)을 통해 접근하기에 각 플랫폼의 특성에 마추는 작업에서 문제가 발생한 것같아요.

쓰다보니 나름의 룰이 정해지는 것 같아요.

**Reply by gyutae100** [reply]
아직 학습이 부족해서, AirBnb 디자인 시스템에서 컴포넌트 기반이 적합하지 않은 이유의 근거를 정확히 이해하지 못하고 있습니다.

말씀해주신, `airbnb는 전세계 사람들이 모든 플랫폼(안드로이드, iOS, 테블릿, 웹 etc)을 통해 접근하기에 각 플랫폼의 특성에 마추는 작업에서 문제가 발생`하는 부분을 파악하도록 하겠습니다.

좋은 답변 감사합니다.

**Reply by peanudge** [reply]
저도 제 생각일뿐이에요. 😅

테블릿 같은 경우, 앱에서 사용하던 디자인을 그대로 사용하면 **어울리지 않는다**같은 느낌이 있을때가 있지않나요? 그런 사례가 떠올라서 그것도 이유겠구나 싶었어요. 즉, 안드로이드 개발자와 iOS 개발자가 추구하는게 다를 수 있는 것처럼 디자인도 거기에 발 맞추려면 파운데이션 수준의 디자인시스템에 대한 니즈가 있을 것 같아요.

**Comment by gyutae100** [comment]
# 2. [FEConf 2022 디자인 시스템, 형태를 넘어서](https://www.youtube.com/watch?v=21eiJc90ggo)

## 높은 유연성과 낮은 제약성을 가진 디자인 시스템의 핵심 가치는 무엇일까?
- 정답은 변화하는 제품에 대응 가능한 높은 생산성이라고 생각합니다.
- 근거
   -  변화하는 제품은 Iterator을 거치며 고객의 요구사항을 반영하거나,
   - 기존의 결정을 번복하거나,
   - 다른 결정을 내리기도 한다.
   - 이런 상황이 발생하는 근본적인 이유는, 모든 요구사항을 사전에 파악하기 불가능하기 때문입니다.
  
## 높은 유연성과 낮은 제약성을 가진 디자인 시스템을 구성하는 요소
1. 높은 유연성
2. 높은 확장성 (유연성과 연관)
3. 낮은 제약성
   - 주요 단점은, 파편화와 인터페이스 사용처에서 커스텀 범위에 대한 혼란 제공
 4. 낮은 통일성 (제약과 연관)
 5. 높은 자유도 (유연성과 연관)
 6. 목적과 의도가 명료한 인터페이스 
    
     - React 예시로, 만약 복합적인 형태와 기능을 제공하는 단일 컴포넌트에서 모든 props의 위계를 관리하는 경우
        - 복합적인 형태는 여러 컴포넌트의 합성을 의미함. 즉 단일 컴포넌트 props로 컴포넌트의 위치, 배치 등을 관리 해야만 한다.
        - 복잡한적인 형태에 대한 기능 또한, 단일 컴포넌트 props로 관리 시, props에 대한 목적과 의도 파악이 쉽지 않으면서 props를 통일성 있게 정의 내리기 어려움.
        - 이러한 개발 방식은 지극히 생산자 중심적이다.
        - 이를 위한 해결책으로, 컴포넌트 분리를 통해서 props가 실제 사용되는 컴포넌트와 바인딩하면 해결할 수 있는 문제이다. 
        - **결론으로, 인터페이스의 목적과 의도는 명료해야 한다.**
7. 디자인 시스템 낯섬 예산 최소화
     - 주요 연관 요소는, 러닝 커브 (또는 진입 장벽), DX
8. 높은 추상화
     - 변경에 '유연하다'는 의미는, 컴포넌트의 스펙이 범용적인 용례를 지원할 수 있어야 한다.
9. 틀린 선택 막아주기
    - 반드시 제약보다 유연함을 추구한다는 것이, 소비자의 실수에 책임을 지지 않는다는 의미는 아니다. 
    - 시스템 레벨에서 표현의 제약은 줄이되, 틀린 표현은 하지 않도록 막아줄 책임은 있다.
10. 접근성
    - 디자인 시스템은 일관된 사용성을 보장해야 한다는 점에서, 접근성 지원은 가능한 소비자가 아닌 디자인 시스템의 역할과 책임이 되어야만 한다.

## 결론
- 위 요소를 만족시킬 수록, 비례해서 `많은 비용`이 필요로 합니다.
- 만약 상황에 맞지 않는 맥락에서 파운데이션 방식이 도입된다면, `오버 엔지니어링` 그이상 그이하도 아닙니다.
- 제약이 낮은 컴포넌트의 사용처 측면에서 단점은 인터페이스 외부에서 추가 로직 구현을 위한 추가적인 비용이 소비된다는 점입니다.
- 이를 위한 해결책으로 `Antd Pro 디자인 시스템` 처럼, 파운데이션 컴포넌트를 결합해서 높은 제약에 의해서 보다 구체성을 띄는 `컴포넌트 디자인 시스템`을 만들어 해결 합니다.
- **이는 파운데이션 디자인 시스템을 통해서 컴포넌트 디자인 시스템을 구축 가능함을 의미합니다. (역은 성립되지 않습니다.)**
- 만약 자원에 의존성 없이 파운데이션 디자인 시스템과 컴포넌트 디자인 시스템을 모두 구현해야 한다면, 파운데이션 디자인 시스템을 먼저 구성 후, 컴포넌트 디자인 시스템을 구성한다면 최적의 디자인 시스템 개발 방법론이라고 판단됩니다.
- **결과적으로 파운데이션 디자인 시스템은 결합을 통해서 제약을 주입할 수 있고, 이를 통해서 컴포넌트 디자인 시스템 구현이 가능하다고 보입니다.**
- **이 경우 굳이 두 개념을 구분할 필요가 없다고 판단할 수 있지만, 한정된 자원 내에서 양자택일의 순간에서는, 생산성 관점에서, `유연한 대처를 중점`시 파운데이션 디자인 시스템을 선택하고, `통일성 중점`시 컴포넌트 디자인 시스템을 선택하면 됩니다.**
- 즉 `적정 비용 측면`에서 파운데이션 디자인 시스템과 컴포넌트 디자인 시스템을 구분 하는 것이 적합하다고 판단합니다.

**Reply by peanudge** [reply]
Headless UI 관심있으니 그것도 한번 같이 고려해서 정리해서 공유해주시면 도움 될 것 같네요. 저는 이번에 처음들어서 완전 높은 자유도?를 제공해줄 것 같다라는 느낌이외에는 감이 잡히진않은 것 같아요.

**Reply by gyutae100** [reply]
# Headless 
## 정리
- Headless의 이점은, 높은 형태의 자유도에 있다고 보입니다.
- 사실 높은 형태가 아닌, Headless는 `형태에 대한 제약`이 아에 없습니다.
- 즉 Headless의 경우 형태가 아에 없으면, 소비자측에서 형태 구현의 역할과 책임을 맡게 됩니다.
- 추가적으로, 구현된 형태와 기능의 연결 또한 소비자의 역할과 책임으로 위임됩니다. (이는 상황에 따라 장점일 수도 있고, 단점일 수도 있음.)
- 좋은 오픈소스 예시는, `react-table`이 있습니다.
- **결론으로 Headless 채택 또한 Trade-off 상, 한정된 자원 속에서 생산성 관점에서 결정할 선택의 문제로 보입니다.**

## 참고

1. (react-table) https://tanstack.com/table/v8/?from=reactTableV7&original=https://react-table-v7.tanstack.com/

2. (react-table의 headless 설명) https://tanstack.com/table/v8/docs/guide/introduction

3. (headless 설명) https://www.merrickchristensen.com/articles/headless-user-interface-components/

**Reply by peanudge** [reply]
그렇다면 아마 Headless만 쓰진 않을 것 같고  패턴화된 기능자체만 healessUI가 담당하게하면 효율적인 용도로 쓸 수 있겠네요.

좋네요.
